flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/lookup/FileStoreLookupFunction.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.FileUtils;
+
+import org.apache.flink.shaded.guava30.com.google.common.primitives.Ints;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.table.store.connector.RocksDBOptions.LRU_CACHE_MAX_ROWS;
+import static org.apache.flink.table.store.file.predicate.PredicateBuilder.transformFieldMapping;
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** A lookup {@link TableFunction} for file store. */
+public class FileStoreLookupFunction extends TableFunction<RowData> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FileStoreLookupFunction.class);
+
+    private final FileStoreTable table;
+    private final List<String> projectFields;
+    private final List<String> joinKeys;
+    @Nullable private final Predicate predicate;
+
+    private transient Duration refreshInterval;
+    private transient File path;
+    private transient RocksDBStateFactory stateFactory;
+    private transient LookupTable lookupTable;
+
+    // timestamp when cache expires
+    private transient long nextLoadTime;
+    private transient TableStreamingReader streamingReader;
+
+    public FileStoreLookupFunction(
+            FileStoreTable table,
+            int[] projection,
+            int[] joinKeyIndex,
+            @Nullable Predicate predicate) {
+        TableSchema schema = table.schema();
+        checkArgument(
+                schema.partitionKeys().isEmpty(), "Currently only support non-partitioned table.");
+        checkArgument(schema.primaryKeys().size() > 0, "Currently only support primary key table.");
+        this.table = table;
+
+        // join keys are based on projection fields
+        this.joinKeys =
+                Arrays.stream(joinKeyIndex)
+                        .mapToObj(i -> schema.fieldNames().get(projection[i]))
+                        .collect(Collectors.toList());
+
+        this.projectFields =
+                Arrays.stream(projection)
+                        .mapToObj(i -> schema.fieldNames().get(i))
+                        .collect(Collectors.toList());
+
+        // add primary keys
+        for (String field : schema.primaryKeys()) {
+            if (!projectFields.contains(field)) {
+                projectFields.add(field);
+            }
+        }
+
+        this.predicate = predicate;
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        super.open(context);
+        String tmpDirectory = getTmpDirectory(context);
```

Comment:
Could the directory be configurable? For the production environment, the temp directory is rejected for users.

Reply:
No, this is not java temp directory, this is work directory of Flink. State will use this directory too.